services/imap/imap.go

```diff
@@ -0,0 +1,365 @@
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/emersion/go-imap"
+	"github.com/emersion/go-imap/client"
+	"github.com/emersion/go-message/mail"
+)
+
+// Client an imap clientor
+type Client struct {
+	Client   *client.Client
+	UserName string
+	Passwd   string
+	Addr     string
+	IsTLS    bool
+	Lock     sync.Mutex
+}
+
+// ClientInitOpt options to init an Client
+type ClientInitOpt struct {
+	Addr     string
+	UserName string
+	Passwd   string
+	IsTLS    bool
+}
+
+// NewImapClient init a imap Client
+func NewImapClient(opt ClientInitOpt) (c *Client, err error) {
+	c = new(Client)
+
+	c.UserName = opt.UserName
+	c.Passwd = opt.Passwd
+	c.Addr = opt.Addr
+	c.IsTLS = opt.IsTLS
+
+	// try login
+	if err = c.Login(); err != nil {
+		return nil, err
+	}
+
+	if err = c.LogOut(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Login login to service
+func (c *Client) Login() error {
+	var err error
+
+	c.Lock.Lock()
+
+	// Connect to server
+	if c.IsTLS {
+		c.Client, err = client.DialTLS(c.Addr, nil)
+	} else {
+		c.Client, err = client.Dial(c.Addr)
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.Client.Login(c.UserName, c.Passwd)
+}
+
+// LogOut LogOut from service
+func (c *Client) LogOut() error {
+	err := c.Client.Logout()
+	c.Client = nil
+	c.Lock.Unlock()
+	return err
+}
+
+// GetUnReadMailIDs get all unread mails
+func (c *Client) GetUnReadMailIDs(mailBox string) ([]uint32, error) {
+	if err := c.Login(); err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := c.LogOut()
+		if err != nil {
+			log.Warn("Imap.Logout", err)
+		}
+	}()
+
+	if len(mailBox) == 0 {
+		mailBox = "INBOX"
+	}
+
+	// Select mail box
+	_, err := c.Client.Select(mailBox, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set search criteria
+	criteria := imap.NewSearchCriteria()
+	criteria.WithoutFlags = []string{imap.SeenFlag}
+	ids, err := c.Client.Search(criteria)
+	if err != nil {
+		return nil, err
+	}
+
+	return ids, err
+}
+
+// Store store status
+func (c *Client) Store(mailBox string, mID uint32, isAdd bool, flags []interface{}) error {
+	if err := c.Login(); err != nil {
+		return err
+	}
+	defer func() {
+		err := c.LogOut()
+		if err != nil {
+			log.Warn("Imap.Logout", err)
+		}
+	}()
+
+	return c.store(mailBox, mID, isAdd, flags)
+}
+
+// store store status without login
+func (c *Client) store(mailBox string, mID uint32, isAdd bool, flags []interface{}) error {
+	if len(mailBox) == 0 {
+		mailBox = "INBOX"
+	}
+
+	// Select INBOX
+	_, err := c.Client.Select(mailBox, false)
+	if err != nil {
+		return err
+	}
+
+	seqSet := new(imap.SeqSet)
+	seqSet.AddNum(mID)
+
+	var opt imap.FlagsOp
+	if isAdd {
+		opt = imap.AddFlags
+	} else {
+		opt = imap.RemoveFlags
+	}
+
+	item := imap.FormatFlagsOp(opt, true)
+
+	return c.Client.Store(seqSet, item, flags, nil)
+}
+
+// DeleteMail delete one mail
+func (c *Client) DeleteMail(mailBox string, mID uint32) error {
```

Comment:
We don't need this method.

Reply:
I have added an optional choice to delete email after read it successly.